### DISCUSSION TITLE
Ability to pass a custom icon to <Aside>

### DISCUSSION THREAD

**Discussion by shubham-padia** [withastro/starlight#2251] [Feature Requests]
### What version of `starlight` are you using?

0.25.4

### What is your idea?

We have multiple instances where we want to show `<Asides>` suggesting keyboard shortcuts. Changing the name of the Aside is possible right now, but we also want to display a keyboard icon on that aside. 

### Why is this feature necessary?

Since the Aside component is relatively small in terms of LOC, it would be possible to just make our own component based on the original implementation and go from there. However, it would be good to have upstream support for it if it meet's the project's criteria.

### Do you have examples of this feature in other projects?

https://zulip.com/help/quote-and-reply

<img width="784" alt="Screenshot 2024-08-22 at 7 49 18 AM" src="https://github.com/user-attachments/assets/c39b7b9a-9cad-4208-9cfa-034a43329f60">

### Participation

- [X] I am willing to submit a pull request for this feature.

**Comment by delucis** [comment]
I think this is a reasonable request. The one tricky part is how to handle this for our `:::note`-style syntax. For the regular Astro `<Aside>` component, adding props is no problem, but we’d want to keep feature parity with the directive syntax too.

**Comment by shubham-padia** [comment]
According to https://github.com/micromark/micromark-extension-directive#syntax, which is what is used in https://github.com/remarkjs/remark-directive?tab=readme-ov-file, we can specify an attribute in this syntax: `::name[label]{attributes}`. 

We can define the extra icons as the iconPath objects defined in [asides.ts](https://github.com/withastro/starlight/blob/6df2f34e81c24e4316a9844a2a94a665457f6a36/packages/starlight/integrations/asides.ts#L112). One option is to define that as part of starlight config object in astro.config.ts, I'm not sure whether that's the best place for it or not. Any suggestion there would be welcome, thanks!

**Reply by delucis** [reply]
The path I’d recommend is to only support the names of Starlight’s built-in icons for now (we can then add missing icons as needed), but that way the syntax can be something like:

```mdx
<Aside type="note" icon="astro" title="Astro wants you to know…">
Contents
</Aside>
```

or

```md
:::note[Astro wants you to know…]{icon="astro"}
Contents
:::
```

I think it should be possible for the `asides.ts` plugin to import the icon paths from [our `Icons.ts` file](https://github.com/withastro/starlight/blob/main/packages/starlight/components/Icons.ts) so this is the same as the icons available in other Starlight components?

**Reply by delucis** [reply]
My only doubt here is that we’ve discussed using https://www.astroicon.dev/ to make customising icons easier, but I’m not sure we’d be able to access that from a remark plugin. So that could complicate things a bit, but we can probably face that problem when/if we come to it in the future.

**Reply by shubham-padia** [reply]
The proposal of using starlight's built-in icons for now sounds good to me. And we can think about astroicon.dev when it comes to it. I'm thinking of starting work on this in a day or two to see if there are additional comments, please let me know if any!

**Reply by delucis** [reply]
No more comments from me — feel free to ask any questions you run into as you work on it or to open an early draft PR if you need feedback 🙌

**Reply by shubham-padia** [reply]
Opened  #2261, please have a look when you get some time, thanks!

**Comment by HiDeoo** [comment]
Starting with [Starlight 0.35.0](https://github.com/withastro/starlight/releases/tag/%40astrojs%2Fstarlight%400.35.0), support for [using any of Starlight’s built-in icons](https://starlight.astro.build/guides/authoring-content/#custom-aside-icons) in asides has been added.

Thanks for your amazing work Shubham!
